File: src/main/java/com/demonwav/mcdev/inspection/IsCancelledInspection.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2016 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.inspection;

import com.demonwav.mcdev.platform.MinecraftModule;

import com.intellij.openapi.module.Module;
import com.intellij.openapi.module.ModuleUtilCore;
import com.intellij.psi.PsiMethodCallExpression;
import com.siyeh.ig.BaseInspection;
import com.siyeh.ig.BaseInspectionVisitor;
import com.siyeh.ig.InspectionGadgetsFix;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;

public class IsCancelledInspection extends BaseInspection {
    @Nls
    @NotNull
    @Override
    public String getDisplayName() {
        return "Useless event is cancelled check";
    }

    @NotNull
    @Override
    protected String buildErrorString(Object... infos) {
        final IsCancelled useless = (IsCancelled) infos[0];
        return useless.getErrorString();
    }

    @Nullable
    @Override
    protected InspectionGadgetsFix buildFix(Object... infos) {
        final IsCancelled useless = (IsCancelled) infos[0];
        return useless.getBuildFix();
    }

    @Override
    public BaseInspectionVisitor buildVisitor() {
        return new BaseInspectionVisitor() {
            @Override
            public void visitMethodCallExpression(PsiMethodCallExpression expression) {
                final Module module = ModuleUtilCore.findModuleForPsiElement(expression);
                if (module == null) {
                    return;
                }

                final MinecraftModule instance = MinecraftModule.getInstance(module);
                if (instance == null) {
                    return;
                }

                Optional<IsCancelled> useless = instance.getModules().stream()
                        .map(m -> m.checkUselessCancelCheck(expression))
                        .filter(i -> i != null)
                        .findAny();
                if (!useless.isPresent()) {
                    return;
                }

                registerMethodCallError(expression, useless.get());
            }
        };
    }
}

File: src/main/kotlin/com/demonwav/mcdev/inspection/IsCancelledInspection.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.inspection

import com.demonwav.mcdev.facet.MinecraftFacet
import com.demonwav.mcdev.util.mapFirstNotNull
import com.intellij.openapi.module.ModuleUtilCore
import com.intellij.psi.PsiMethodCallExpression
import com.siyeh.ig.BaseInspection
import com.siyeh.ig.BaseInspectionVisitor
import com.siyeh.ig.InspectionGadgetsFix
import org.jetbrains.annotations.Nls

class IsCancelledInspection : BaseInspection() {
    @Nls
    override fun getDisplayName() = "Useless event is cancelled check"

    override fun buildErrorString(vararg infos: Any): String {
        val useless = infos[0] as IsCancelled
        return useless.errorString
    }

    override fun buildFix(vararg infos: Any): InspectionGadgetsFix? {
        val useless = infos[0] as? IsCancelled
        return useless?.buildFix
    }

    override fun buildVisitor(): BaseInspectionVisitor {
        return object : BaseInspectionVisitor() {
            override fun visitMethodCallExpression(expression: PsiMethodCallExpression) {
                val module = ModuleUtilCore.findModuleForPsiElement(expression) ?: return

                val instance = MinecraftFacet.getInstance(module) ?: return

                val useless = instance.modules.mapFirstNotNull { m -> m.checkUselessCancelCheck(expression) } ?: return

                registerMethodCallError(expression, useless)
            }
        }
    }
}
